### PR TITLE
[Issue-709] Leverage `DeliveryGuarantee` in new sink API

### LINKED ISSUE #709: Use `DeliveryGuarantee` enum as sink configuration
**Problem description**
Flink has already provided a enum `DeliveryGuarantee` to describe the delivery guarantee that our sink uses. We should leverage it instead of the current `PravegaWriterMode` enum.

**Problem location**
`PravegaWriterMode`, Sink

**Suggestions for an improvement**

### PR DESCRIPTION
**Pull request by fyang86**
Signed-off-by: Fan, Yang <fan.yang5@emc.com>

**Change log description**
Flink has already provided a enum `DeliveryGuarantee` to describe the delivery guarantee. We will use it instead of `PravegaWriterMode` in our new sink API introduced in #460 

**Purpose of the change**
Fix #709 

**What the code does**
Replace `PravegaWriterMode` with `DeliveryGuarantee` in new sink API.

**How to verify it**
`./gradlew clean build` should pass

### CODE REVIEW

File: src/main/java/io/pravega/connectors/flink/sink/PravegaEventWriter.java [hunk 1]
```diff
@@ -77,7 +77,7 @@
     private final Stream stream;
 
     // The sink's mode of operation. This is used to provide different guarantees for the written events.
-    private final PravegaWriterMode writerMode;
+    private final DeliveryGuarantee writerMode;
```
Comment (crazyzhou) [inline src/main/java/io/pravega/connectors/flink/sink/PravegaEventWriter.java:80]: The variable name can change as well, just call it `deliveryGuarantee`, same to the helper methods

**Review by crazyzhou — CHANGES_REQUESTED** [review]

**Review by crazyzhou — APPROVED** [review]
LGTM

File: src/main/java/io/pravega/connectors/flink/sink/PravegaEventWriter.java [hunk 2]
```diff
@@ -228,8 +228,8 @@ public void close() throws Exception {
     }
 
     @VisibleForTesting
```
Comment (crazyzhou) [inline src/main/java/io/pravega/connectors/flink/sink/PravegaEventWriter.java:230]: Here we mark it `VisibleForTesting`, but I'm not seeing it being used in tests. The coverage is also below par, needs to add some more test to cover

**Review by crazyzhou — CHANGES_REQUESTED** [review]

File: src/main/java/io/pravega/connectors/flink/sink/PravegaSinkBuilder.java [hunk 3]
```diff
@@ -81,20 +81,20 @@ public PravegaSinkBuilder<T> forStream(final Stream stream) {
...
-     * Sets the writer mode to provide at-least-once or exactly-once guarantees.
+     * Sets the delivery guarantee to provide at-least-once or exactly-once guarantees.
      *
-     * @param writerMode The writer mode of {@code BEST_EFFORT}, {@code ATLEAST_ONCE}, or {@code EXACTLY_ONCE}.
+     * @param deliveryGuarantee The delivery guarantee of {@code NONE}, {@code AT_LEAST_ONCE}, or {@code EXACTLY_ONCE}.
      * @return A builder to configure and create a sink.
      */
-    public PravegaSinkBuilder<T> withWriterMode(PravegaWriterMode writerMode) {
-        this.writerMode = writerMode;
+    public PravegaSinkBuilder<T> withDeliveryGuarantee(DeliveryGuarantee deliveryGuarantee) {
```
Comment (crazyzhou) [inline src/main/java/io/pravega/connectors/flink/sink/PravegaSinkBuilder.java:89]: We still need `withWriterMode` method to keep the compatibility.

**Review by crazyzhou — CHANGES_REQUESTED** [review]

File: src/main/java/io/pravega/connectors/flink/sink/PravegaSinkBuilder.java (hunk 3 above)
Comment (fyang86) [reply src/main/java/io/pravega/connectors/flink/sink/PravegaSinkBuilder.java:90]: fixed

File: src/main/java/io/pravega/connectors/flink/sink/PravegaSinkBuilder.java [hunk 4]
```diff
@@ -81,20 +82,42 @@ public PravegaSinkBuilder<T> forStream(final Stream stream) {
...
+        this.deliveryGuarantee = deliveryGuarantee;
+        return this;
+    }
+
+    /**
+     * Sets the delivery guarantee from writer mode to provide at-least-once or exactly-once guarantees.
+     * This method is to keep the compatibility with the old {@code PravegaWriterMode} enum that will be deprecated soon.
      *
      * @param writerMode The writer mode of {@code BEST_EFFORT}, {@code ATLEAST_ONCE}, or {@code EXACTLY_ONCE}.
      * @return A builder to configure and create a sink.
```
Comment (crazyzhou) [inline src/main/java/io/pravega/connectors/flink/sink/PravegaSinkBuilder.java:100]: Add a deprecation notice here, suggesting user to switch to `withDeliveryGuarantee` method

File: src/main/java/io/pravega/connectors/flink/sink/PravegaSinkBuilder.java [hunk 5]
```diff
@@ -94,11 +94,12 @@ public PravegaSinkBuilder<T> withDeliveryGuarantee(DeliveryGuarantee deliveryGua
 
     /**
      * Sets the delivery guarantee from writer mode to provide at-least-once or exactly-once guarantees.
-     * This method is to keep the compatibility with the old {@code PravegaWriterMode} enum that will be deprecated soon.
+     * @deprecated This method is to keep the compatibility with the old {@code PravegaWriterMode} enum that will be deprecated soon.
```
Comment (crazyzhou) [inline src/main/java/io/pravega/connectors/flink/sink/PravegaSinkBuilder.java:97]:
```suggestion
     * @deprecated use {@link withDeliveryGuarantee} instead
```

**Review by crazyzhou — APPROVED** [review]
LGTM, please resolve the conflicts

**Pull request merged**